AX: Fix for accessibility/password-selected-text-range.html in ITM.

AndresGonzalezApple opened the pull request:
#### fc5dad73c18c845cf725213118e9de41a1691eb8
<pre>
AX: Fix for accessibility/password-selected-text-range.html in ITM.
<a href="https://bugs.webkit.org/show_bug.cgi?id=292031">https://bugs.webkit.org/show_bug.cgi?id=292031</a>
&lt;<a href="https://rdar.apple.com/problem/149980238">rdar://problem/149980238</a>&gt;

Reviewed by Tyler Wilcock.

SelectedTextChanged notifications need to be exposed for secure text fields since AX clients depend on these notifications to retrieve the correct selection range. This is one of the problems that is causing issues with entering passwords on the web with some assistive technologies.

To expose the SelectedTextchanged notifications, this patch adds a mechanism of queueing text changes that keeps all the necessary data, encapsulated in the AXTextChangeContext struct.
These changes also required changing the input-replacevalue-userinfo.html since we now expose more accurate values for the password fields. this prompted a complete re-write of this test to bring it to our current standards and correct several issues with the test itself.

* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/mac/input-replacevalue-userinfo-expected.txt:
* LayoutTests/accessibility/mac/input-replacevalue-userinfo.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::showIntent):
(WebCore::AXObjectCache::postTextStateChangeNotification):
(WebCore::AXObjectCache::postTextReplacementNotification):
(WebCore::AXObjectCache::postTextReplacementNotificationForTextControl):
(WebCore::secureContext):
(WebCore::AXObjectCache::enqueuePasswordNotification):
(WebCore::AXObjectCache::passwordNotificationTimerFired):
(WebCore::AXObjectCache::onSelectedTextChanged):
(WebCore::AXObjectCache::textChangeForEditType):
(WebCore::AXObjectCache::passwordNotificationPostTimerFired): Deleted.
(WebCore::AXObjectCache::enqueuePasswordValueChangeNotification): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextStateChangeIntent.h:
(WebCore::AXTextStateChangeIntent::AXTextStateChangeIntent):
* Source/WebCore/accessibility/atspi/AXObjectCacheAtspi.cpp:
(WebCore::AXObjectCache::postTextStateChangePlatformNotification):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(platformEditTypeForWebCoreEditType):
(WebCore::AXObjectCache::onSelectedTextChanged): Deleted.

Canonical link: <a href="https://commits.webkit.org/294435@main">https://commits.webkit.org/294435@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39d1afa38543da1f8c0e94f052ccb18fc68a1bde

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21461 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11777 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106951 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52427 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29966 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77510 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34530 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104800 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16820 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91921 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57848 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16646 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9939 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51778 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86501 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10016 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109308 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28926 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21304 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86489 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29287 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88122 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86060 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21905 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30812 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8532 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23096 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28854 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34144 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28665 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31988 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30224 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2436,43 +2427,151 @@ void AXObjectCache::postTextReplacementNotificationForTextControl(HTMLTextFormCo
     if (!observableObject) {
         ASSERT_NOT_REACHED();
-        // return true even though the enqueue didn't happen because this is a password field and caller shouldn't post a notification
+        // Return true even though the enqueue didn't happen because this is a password field and caller shouldn't post a notification unless is queued.
```
Thanks for fixing the sentence-casing on this. You wrote "unless is queued" — should it be "unless it is queued"?

AndresGonzalezApple replied:
Fixed.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2436,43 +2427,151 @@ void AXObjectCache::postTextReplacementNotificationForTextControl(HTMLTextFormCo
+        };
+    }
+    updateIsolatedTree(notifications);
 #endif
```
Seems like we made the block inside this `#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)` much larger — could be nice to have annotate this `#endif` with `// ENABLE(ACCESSIBILITY_ISOLATED_TREE)`?

AndresGonzalezApple replied:
Fixed.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2436,43 +2427,151 @@ void AXObjectCache::postTextReplacementNotificationForTextControl(HTMLTextFormCo
+            m_selectedTextRangeTimer.restart();
+    }
+
+    if (auto tree = AXIsolatedTree::treeForPageID(m_pageID)) {
```
I realize this was `auto` before your PR, but I think we've leaned towards explicitly specifying smart pointer types (`RefPtr` here) rather than using `auto`, to make the lifetime semantics of the pointed-to object obvious.

AndresGonzalezApple replied:
Fixed.

hoffmanjoshua commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2436,43 +2427,151 @@ void AXObjectCache::postTextReplacementNotificationForTextControl(HTMLTextFormCo
+
+    // FIXME: Add a better way to retrieve the maskingCharacter for secure fields.
+    String value = object.secureFieldValue();
+    auto maskingCharacter = value.length() ? value[0] : bullet;
```
Is it new behavior to use the first character of the value as the masking character? I see that we are using it to replace the deleted/inserted text in the context—will that cause any confusion for users?

AndresGonzalezApple replied:
Only for secure fields. Both the rendition to the screen and AT should show masking characters, not the real characters.

AndresGonzalezApple committed:
AX: Fix for accessibility/password-selected-text-range.html in ITM.
https://bugs.webkit.org/show_bug.cgi?id=292031
<rdar://problem/149980238>

Reviewed by Tyler Wilcock.

SelectedTextChanged notifications need to be exposed for secure text fields since AX clients depend on these notifications to retrieve the correct selection range. This is one of the problems that is causing issues with entering passwords on the web with some assistive technologies.

To expose the SelectedTextchanged notifications, this patch adds a mechanism of queueing text changes that keeps all the necessary data, encapsulated in the AXTextChangeContext struct.
These changes also required changing the input-replacevalue-userinfo.html since we now expose more accurate values for the password fields. this prompted a complete re-write of this test to bring it to our current standards and correct several issues with the test itself.

* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/mac/input-replacevalue-userinfo-expected.txt:
* LayoutTests/accessibility/mac/input-replacevalue-userinfo.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::showIntent):
(WebCore::AXObjectCache::postTextStateChangeNotification):
(WebCore::AXObjectCache::postTextReplacementNotification):
(WebCore::AXObjectCache::postTextReplacementNotificationForTextControl):
(WebCore::secureContext):
(WebCore::AXObjectCache::enqueuePasswordNotification):
(WebCore::AXObjectCache::passwordNotificationTimerFired):
(WebCore::AXObjectCache::onSelectedTextChanged):
(WebCore::AXObjectCache::textChangeForEditType):
(WebCore::AXObjectCache::passwordNotificationPostTimerFired): Deleted.
(WebCore::AXObjectCache::enqueuePasswordValueChangeNotification): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextStateChangeIntent.h:
(WebCore::AXTextStateChangeIntent::AXTextStateChangeIntent):
* Source/WebCore/accessibility/atspi/AXObjectCacheAtspi.cpp:
(WebCore::AXObjectCache::postTextStateChangePlatformNotification):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(platformEditTypeForWebCoreEditType):
(WebCore::AXObjectCache::onSelectedTextChanged): Deleted.

Canonical link: https://commits.webkit.org/294435@main
